### DISCUSSION TITLE
请问下性能测试的相关细节

### DISCUSSION THREAD

**Discussion by yuyyi51** [Tencent/tquic#197] [Q&A (问题和回答)] [answered]
https://tquic.net/zh/docs/further_readings/benchmark/ 这里提到对比的几个其它QUIC库都使用了项目自带的示例程序，想了解下tquic这边也是使用示例程序测试的吗？我注意像是lsquic底层支持了sendmmsg，而tquic这边的示例程序都是收/发单个包，这块有评估过会对性能有多少影响吗？

**Comment by glime-ai** [comment, marked as answer]
是的，根据提供的信息，TQUIC的性能测试也是使用了示例程序进行的。在性能测试中，各服务端示例程序的应用逻辑被保持足够简单，以便尽量真实反映QUIC协议栈的处理开销。例如，LSQUIC服务端程序使用正则表达式从请求中提取文件路径，这会引入一定的性能开销。因此，为了更准确地反映QUIC协议栈的性能，将正则表达式匹配替换为了简单的字符串匹配。

关于sendmmsg的使用，确实，一些QUIC实现（如LSQUIC）支持sendmmsg来进行批量数据包发送，这可以减少系统调用的次数，从而提高网络I/O效率。TQUIC在示例程序中主要展示了收/发单个包的场景，这主要是为了简化示例的复杂度，使得用户更容易理解和使用。至于sendmmsg对性能的具体影响，这取决于多种因素，包括网络条件、数据包大小、发送频率等。在高负载或高性能要求的场景下，使用sendmmsg等技术进行批量处理确实能够带来性能上的提升。

TQUIC的设计和实现考虑了高性能和高效率，包括支持多路复用、零RTT连接建立、改进的拥塞控制算法等。虽然在示例程序中没有直接展示sendmmsg的使用，但TQUIC的架构和设计允许在需要时集成此类高效的网络I/O操作。

如果您对TQUIC在特定场景下的性能表现有进一步的兴趣，建议进行定制化的性能测试，以评估sendmmsg等技术在您的应用场景中的实际影响。

**Reply by iyangsj** [reply]
为尽可能在一致条件下比较协议栈本身开销，各示例程序都按非批量收发模式测试。
实际使用时，建议采用性能更优的批量收发模式
